source/http2_stream_manager.c

```diff
@@ -0,0 +1,801 @@
+                    "acquisition:%p cannot find any connection to use.",
+                    (void *)pending_acquisition);
+                break;
+            } else {
+                /* found connection for the request. Move it to pending make requests and update the count */
+                aws_linked_list_push_back(&work->pending_make_requests, node);
+                --stream_manager->synced_data.pending_acquisition_count; /* Length of the pending_acquisitions */
+                ++stream_manager->synced_data.scheduled_stream_count;
+            }
+        }
+
+        /* Step 2: Check for new connections needed */
+        if (stream_manager->synced_data.pending_acquisition_count) {
+            size_t num_connections_needed = stream_manager->synced_data.pending_acquisition_count / UINT32_MAX + 1;
+            work->new_connections = num_connections_needed - stream_manager->synced_data.connections_acquiring;
+            stream_manager->synced_data.connections_acquiring += work->new_connections;
+            STREAM_MANAGER_LOGF(
+                TRACE,
+                stream_manager,
+                "number of acquisition that waiting for connections to use=%zu. connection acquiring=%zu",
+                stream_manager->synced_data.pending_acquisition_count,
+                stream_manager->synced_data.connections_acquiring);
+        }
+
+    } else {
+        STREAM_MANAGER_LOG(DEBUG, stream_manager, "Shutting down");
+        /* Fail all the acquisitions */
+        while (!aws_linked_list_empty(&stream_manager->synced_data.pending_acquisitions)) {
+            /* TODO: how to invoke all those callbacks async? */
+            struct aws_linked_list_node *node =
+                aws_linked_list_pop_back(&stream_manager->synced_data.pending_acquisitions);
+            struct aws_h2_sm_pending_stream_acquisition *pending_acquisition =
+                AWS_CONTAINER_OF(node, struct aws_h2_sm_pending_stream_acquisition, node);
+            /* Make sure no connection assigned. */
+            AWS_ASSERT(pending_acquisition->sm_connection == NULL);
+            if (pending_acquisition->callback) {
+                pending_acquisition->callback(
+                    NULL, AWS_ERROR_HTTP_STREAM_MANAGER_SHUTTING_DOWN, pending_acquisition->user_data);
+            }
+            STREAM_MANAGER_LOGF(
+                DEBUG,
+                stream_manager,
+                "acquisition:%p failed due to stream manager shutting down",
+                (void *)pending_acquisition);
+            s_pending_stream_acquisition_destroy(pending_acquisition);
+        }
+        stream_manager->synced_data.pending_acquisition_count = 0;
+        /* For the stream opening, we just waiting for them to complete. Once all the streams completed, stream manager
+         * can die. */
+        /* Step 3: Check should destroy the stream manager or not. */
+        work->should_destroy_manager = s_aws_http2_stream_manager_should_destroy_synced(stream_manager);
```

Comment:
you're absolutely sure this can only be set once ... right?

Reply:
yes